### ISSUE TITLE
Loss curves for Gemma 7B distributed full finetune are not reproducible

### ISSUE THREAD

**Issue by ebsmothers** [pytorch/torchtune#1303]
When I run the same command multiple times, I get different loss curves:

E.g. 

```
tune run --nnodes 1 --nproc_per_node 2 full_finetune_distributed --config gemma/7B_full
```

<img width="1633" alt="Screenshot 2024-08-10 at 1 49 35 PM" src="https://github.com/user-attachments/assets/21a65363-b9b0-4bcd-94c0-8fda5a965040">

This is in contrast to e.g. our Llama3 recipe, where the results are identical (as one would expect):
<img width="1808" alt="Screenshot 2024-08-10 at 12 37 10 PM" src="https://github.com/user-attachments/assets/d0e6eb5e-af0e-4996-9223-07a2a78aec1d">

Note that for the Gemma 7B LoRA distributed recipe, the loss curves do seem to be reproducible:

<img width="1619" alt="Screenshot 2024-08-10 at 1 19 53 PM" src="https://github.com/user-attachments/assets/df50222c-f4f4-4bba-ad8f-3bde97d53169">

Even for Gemma 2B full finetune the losses look the same
<img width="1629" alt="Screenshot 2024-08-10 at 1 20 45 PM" src="https://github.com/user-attachments/assets/2d1eb0df-5b3f-424f-885b-61f808f51067">

Note that there is no actual single-device fine-tuning recipe for Gemma 7B. But I hacked one together ([config](https://gist.github.com/ebsmothers/1e0121bcb01a779495033adef56090a8)) and there is a small but noticeable difference there as well:

<img width="1608" alt="Screenshot 2024-08-10 at 2 00 13 PM" src="https://github.com/user-attachments/assets/29152a19-2052-4a2f-91ea-7c58d920a48e">

However it does appear to be worse with FSDP enabled.
